packages/common/src/viz/sqlRunner.ts

```diff
@@ -0,0 +1,238 @@
+import { DimensionType } from '../types/field';
+import {
+    type BarChartData,
+    type ResultsTransformerBase,
+    type RowData,
+} from './base';
+
+export type SqlColumn = {
+    id: string;
+    type: DimensionType;
+};
+
+type XLayoutOptions = {
+    columnId: string;
+    timeGrainOptions: string[];
+};
+
+type YLayoutOptions = {
+    columnId: string;
+    aggregationOptions: string[];
+};
+
+type SeriesLayoutOptions = {
+    columnId: string;
+};
+
+export type SqlTransformBarChartConfig = {
+    x: {
+        columnId: string;
+        timeGrain: string;
+    };
+    y: {
+        columnId: string;
+        aggregation: string;
+        sort: string | null;
+    }[];
+    groupBy: {
+        columnId: string | null;
+    };
+};
+export type DuckDBSqlFunction = (
+    sql: string,
+    rowData: RowData[],
+) => Promise<RowData[]>;
+
+type GetPivotedResultsArgs = {
+    rows: RowData[];
+    valuesSql: string[];
+    pivotsSql: string[];
+    groupByColumns: string[];
+    sortsSql: string[];
+    duckDBSqlFunction: DuckDBSqlFunction;
+};
+
+export const getPivotedResults = async ({
+    rows,
+    valuesSql,
+    pivotsSql,
+    groupByColumns,
+    sortsSql,
+    duckDBSqlFunction,
+}: GetPivotedResultsArgs) => {
+    const pivotOnSql = pivotsSql.join(', ');
+    const pivotValuesSql = valuesSql.join(', ');
+
+    let query = 'PIVOT results_data';
+    if (pivotsSql.length > 0) {
+        query += ` ON ${pivotOnSql}`;
+    }
+    if (valuesSql.length > 0) {
+        query += ` USING ${pivotValuesSql}`;
+    } else {
+        return {
+            results: [],
+            columns: [],
+        };
+    }
+    if (groupByColumns.length > 0) {
+        query += ` GROUP BY ${groupByColumns.join(', ')}`;
+    }
+    if (sortsSql.length > 0) {
+        query += ` ORDER BY ${sortsSql.join(', ')}`;
+    }
+
+    const pivoted = await duckDBSqlFunction(query, rows);
+    const fieldNames = Object.keys(pivoted[0]);
+
+    return {
+        results: pivoted,
+        indexColumns: groupByColumns,
+        valueColumns: fieldNames.filter(
+            (field) => !groupByColumns.includes(field),
+        ),
+    };
+};
+
+type SqlRunnerResultsTransformerDeps = {
+    duckDBSqlFunction: DuckDBSqlFunction;
+    rows: RowData[];
+    columns: SqlColumn[];
+};
+
+export class SqlRunnerResultsTransformer2
+    implements ResultsTransformerBase<SqlTransformBarChartConfig>
+{
+    private readonly duckDBSqlFunction: DuckDBSqlFunction;
```

Comment:
Do we think this will be a more generic resutlsTransformationFuntion? It could be duck DB or a warehouse call or just an in-memory function? Or if we think we have duckDB available everywhere, this is fine.